sw/device/tests/sram_ctrl_smoketest.c

```diff
@@ -0,0 +1,87 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+static const uint32_t random_data[16] = {
+    0x6b4abfae, 0x63bdb6e7, 0x87f99b1a, 0xa214dffe, 0xb12291f9, 0xd0cd1abe,
+    0x5c95e716, 0xe887aab1, 0x307f6ef9, 0x6f5c0464, 0x5882279d, 0x44c19574,
+    0x1bd20079, 0xf8250ead, 0x4bf362a4, 0xad41437d};
+
+// Buffer to allow the compiler to allocate a safe area in Main SRAM where
+// we can do the write/read test without the risk of clobbering data
+// used by the program.
+__attribute__((
+    section(".data"))) static volatile const uint32_t sram_main_buffer[16] = {
+    0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
+    0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000,
+    0x00000000, 0x00000000, 0x00000000, 0x00000000};
+
+// Write / Read small chunks of data to SRAM to test
+// basic functionality of sram_ctrl
+static void write_read_check(mmio_region_t base_addr_region,
+                             const char *sram_name) {
+  for (int i = 0; i < 16; ++i) {
+    mmio_region_write32(base_addr_region, i * sizeof(uint32_t), random_data[i]);
+  }
+  for (int i = 0; i < 16; ++i) {
+    uint32_t rw_data_32 =
+        mmio_region_read32(base_addr_region, i * sizeof(uint32_t));
+    CHECK(rw_data_32 == random_data[i],
+          "Memory Write/Read Mismatch for %s, index %d, data read = %8x "
+          "data_expected = %8x.",
+          sram_name, i, rw_data_32, random_data[i]);
+  }
+}
```

Comment:
Nit: These two loops probably could be combined, so it would be:
WRITE - READ - COMPARE
In a single loop.